File: ThreemaFramework/Message/Sender/MessageForwarder.m
//  _____ _
// |_   _| |_  _ _ ___ ___ _ __  __ _
//   | | | ' \| '_/ -_) -_) '  \/ _` |_
//   |_| |_||_|_| \___\___|_|_|_\__,_(_)
//
// Threema iOS Client
// Copyright (c) 2015-2022 Threema GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

#import "MessageForwarder.h"
#import "EntityCreator.h"
#import "EntityFetcher.h"
#import "ThreemaFramework/ThreemaFramework-Swift.h"
#import "MessageSender.h"
#import "TextMessage.h"
#import "ImageMessageEntity.h"
#import "VideoMessageEntity.h"
#import "AudioMessageEntity.h"
#import "LocationMessage.h"
#import "FileMessageSender.h"
#import "UTIConverter.h"
#import <ThreemaFramework/ThreemaFramework-Swift.h>

#ifdef DEBUG
  static const DDLogLevel ddLogLevel = DDLogLevelVerbose;
#else
  static const DDLogLevel ddLogLevel = DDLogLevelWarning;
#endif
@implementation MessageForwarder

+ (void)forwardMessage:(BaseMessage *)message toContact:(Contact *)contact
{
    EntityManager *entityManager = [[EntityManager alloc] init];
    Conversation *conversation = [entityManager conversationForContact:contact createIfNotExisting:YES];
    
    [self forwardMessage:message toConversation:conversation];
}

+ (void)forwardMessage:(BaseMessage *)message toGroup:(Group *)group
{

    EntityManager *entityManager = [[EntityManager alloc] init];
    Conversation *conversation = [[entityManager entityFetcher] conversationForGroupId:group.groupID creator:group.groupCreatorIdentity];

    [self forwardMessage:message toConversation:conversation];
}

+ (void)forwardMessage:(BaseMessage *)message toConversation:(Conversation *)conversation
{
    if (conversation == nil) {
        DDLogError(@"MessageForwarder: Cannot send message to nil conversation.");
        return;
    }
    
    if ([message isKindOfClass: [TextMessage class]]) {
        TextMessage *textMessage = (TextMessage *)message;
        [MessageSender sendMessage:textMessage.text inConversation:conversation quickReply:NO requestId:nil onCompletion:^(BaseMessage *message) {
            ;//nop
        }];
    } else if ([message isKindOfClass: [AudioMessageEntity class]]) {
        AudioMessageEntity *audioMessageEntity = (AudioMessageEntity *)message;
        NSData *data = [audioMessageEntity.audio.data copy];

        URLSenderItem *item = [URLSenderItem itemWithData:data
                                                 fileName:@"AudioMessage"
                                                     type:UTTYPE_AUDIO
                                               renderType:@1
                                               sendAsFile:true];

        FileMessageSender *sender = [[FileMessageSender alloc] init];
        [sender sendItem:item inConversation:conversation requestId:nil];
        
    } else if ([message isKindOfClass: [ImageMessageEntity class]]) {
        ImageMessageEntity *imageMessageEntity = (ImageMessageEntity *)message;
        ImageURLSenderItemCreator *imageSender = [[ImageURLSenderItemCreator alloc] init];
        URLSenderItem *item = [imageSender senderItemFromImage:imageMessageEntity.image.uiImage];

        FileMessageSender *sender = [[FileMessageSender alloc] init];
        [sender sendItem:item inConversation:conversation];
    } else if ([message isKindOfClass: [VideoMessageEntity class]]) {
        VideoMessageEntity *videoMessageEntity = (VideoMessageEntity *)message;
        
        VideoURLSenderItemCreator *senderCreator = [[VideoURLSenderItemCreator alloc] init];
        NSURL *videoURL = [VideoURLSenderItemCreator writeToTemporaryDirectoryWithData:videoMessageEntity.video.data];
        if (videoURL == nil) {
            DDLogError(@"VideoURL was nil.");
            return;
        }
        
        URLSenderItem *senderItem = [senderCreator senderItemFrom:videoURL];
        
        FileMessageSender *sender = [[FileMessageSender alloc] init];
        [sender sendItem:senderItem inConversation:conversation requestId:nil];
        (void)[VideoURLSenderItemCreator cleanTemporaryDirectory];
        
    } else if ([message isKindOfClass: [LocationMessage class]]) {
        LocationMessage *locationMessage = (LocationMessage *)message;
        
        CLLocationCoordinate2D coordinates = CLLocationCoordinate2DMake(locationMessage.latitude.doubleValue, locationMessage.longitude.doubleValue);
        double acurracy = locationMessage.accuracy.doubleValue;
        [MessageSender sendLocation:coordinates accuracy:acurracy poiName:locationMessage.poiName poiAddress:locationMessage.poiAddress inConversation:conversation onCompletion:^(NSData *messageId) {
            ;//nop
        }];
    }
}

@end

File: ThreemaFramework/Message/Sender/MessageForwarder.swift
import CocoaLumberjackSwift
import CoreLocation
import Foundation

@MainActor
public final class MessageForwarder {

    // MARK: - Public types

    public enum AdditionalContent {
        case caption(String)
        case text(String)
    }

    private let businessInjector = BusinessInjector.ui

    // MARK: - Lifecycle

    public init() { }
    
    // MARK: - Forwarding

    /// Forwards a given base message to a conversation and adds additional text as caption or additional text message.
    /// - Parameters:
    ///   - message: Base message to be forwarded.
    ///   - conversation: ConversationEntity which message should be forwarded to.
    ///   - additionalContent: Additional content to be send as caption or as normal text message.
    ///
    ///   - Note: Caption content can only be sent for for file messages.
    public func forward(
        _ message: BaseMessageEntity,
        to conversation: ConversationEntity,
        sendAsFile: Bool,
        additionalContent: MessageForwarder.AdditionalContent?,
    ) {
        switch message {
        case let textMessage as TextMessageEntity:
            Task {
                await businessInjector.messageSender.sendTextMessage(containing: textMessage.text, in: conversation)
                if case let .text(additionalText) = additionalContent {
                    await businessInjector.messageSender.sendTextMessage(containing: additionalText, in: conversation)
                }
            }

        case let locationMessage as LocationMessageEntity:
            Task {
                let coordinates = CLLocationCoordinate2DMake(
                    locationMessage.latitude.doubleValue,
                    locationMessage.longitude.doubleValue
                )
                let accuracy = locationMessage.accuracy?.doubleValue ?? 0.0

                await businessInjector.messageSender.sendLocationMessage(
                    coordinates: coordinates,
                    accuracy: accuracy,
                    poiName: locationMessage.poiName,
                    poiAddress: locationMessage.poiAddress,
                    in: conversation
                )

                if case let .text(additionalText) = additionalContent {
                    await businessInjector.messageSender.sendTextMessage(containing: additionalText, in: conversation)
                }
            }

        case let fileMessage as FileMessageEntity:
            
            guard let item = URLSenderItem(
                data: fileMessage.data?.data,
                fileName: fileMessage.fileName,
                type: fileMessage.blobUTTypeIdentifier,
                renderType: sendAsFile ? 0 : fileMessage.type,
                sendAsFile: sendAsFile
            ) else {
                DDLogError("[MessageForwarder] Could not create URLSenderItem.")
                return
            }

            if case let .caption(caption) = additionalContent {
                item.caption = caption
            }

            if let duration = fileMessage.duration {
                item.duration = duration as NSNumber
            }
            
            Task {
                do {
                    try await businessInjector.messageSender.sendBlobMessage(for: item, in: conversation.objectID)
                    if case let .text(additionalText) = additionalContent {
                        await businessInjector.messageSender.sendTextMessage(
                            containing: additionalText,
                            in: conversation
                        )
                    }
                }
                catch {
                    DDLogError("[MessageForwarder] Could not send sender item, error: \(error)")
                }
            }

        case let audioMessage as AudioMessageEntity:
            guard let audioDataEntity = audioMessage.audio,
                  let item = URLSenderItem(
                      data: audioDataEntity.data,
                      fileName: audioDataEntity.getFilename(),
                      type: UTType.audio.identifier,
                      renderType: sendAsFile ? 0 : 1,
                      sendAsFile: sendAsFile
                  ) else {
                DDLogError("[MessageForwarder] Could not create URLSenderItem.")
                return
            }
            
            Task {
                do {
                    try await businessInjector.messageSender.sendBlobMessage(for: item, in: conversation.objectID)
                    if case let .text(additionalText) = additionalContent {
                        await businessInjector.messageSender.sendTextMessage(
                            containing: additionalText,
                            in: conversation
                        )
                    }
                }
                catch {
                    DDLogError("[MessageForwarder] Could not send sender item, error: \(error)")
                }
            }

        case let imageMessage as ImageMessageEntity:
            guard let image = imageMessage.image,
                  let item = URLSenderItem(
                      data: image.data,
                      fileName: image.getFilename(),
                      type: imageMessage.blobUTTypeIdentifier,
                      renderType: sendAsFile ? 0 : 1,
                      sendAsFile: sendAsFile
                  ) else {
                DDLogError("[MessageForwarder] Could not create URLSenderItem.")
                return
            }
            
            if case let .caption(caption) = additionalContent {
                item.caption = caption
            }

            Task {
                do {
                    try await businessInjector.messageSender.sendBlobMessage(for: item, in: conversation.objectID)
                    if case let .text(additionalText) = additionalContent {
                        await businessInjector.messageSender.sendTextMessage(
                            containing: additionalText,
                            in: conversation
                        )
                    }
                }
                catch {
                    DDLogError("[MessageForwarder] Could not send sender item, error: \(error)")
                }
            }
            
        case let videoMessage as VideoMessageEntity:
            guard let video = videoMessage.video,
                  let item = URLSenderItem(
                      data: video.data,
                      fileName: video.getFilename(),
                      type: videoMessage.blobUTTypeIdentifier,
                      renderType: sendAsFile ? 0 : 1,
                      sendAsFile: sendAsFile
                  ) else {
                DDLogError("[MessageForwarder] Could not create URLSenderItem.")
                return
            }
            
            if case let .caption(caption) = additionalContent {
                item.caption = caption
            }

            Task {
                do {
                    try await businessInjector.messageSender.sendBlobMessage(for: item, in: conversation.objectID)
                    if case let .text(additionalText) = additionalContent {
                        await businessInjector.messageSender.sendTextMessage(
                            containing: additionalText,
                            in: conversation
                        )
                    }
                }
                catch {
                    DDLogError("[MessageForwarder] Could not send sender item, error: \(error)")
                }
            }
            
        default:
            DDLogError("[MessageForwarder] Unsupported message type received.")
        }
    }
}
